st2common/st2common/util/execution_queue_db.py

```diff
@@ -0,0 +1,94 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+from mongoengine import ValidationError
+
+from st2common import log as logging
+from st2common.util import date
+from st2common.exceptions.db import StackStormDBObjectNotFoundError
+from st2common.persistence.execution_queue import ExecutionQueue
+from st2common.models.db.execution_queue import ExecutionQueueDB
+from st2common.services import coordination
+
+LOG = logging.getLogger(__name__)
+
+
+__all__ = [
+    'get_execution_request_by_id',
+    'pop_next_execution'
+]
+
+
+def get_execution_request_by_id(action_id):
+    """
+        Get Action by id.
+
+        On error, raise StackStormDBObjectNotFoundError
+    """
+    action = None
+
+    try:
+        action = ExecutionQueue.get_by_id(action_id)
+    except (ValueError, ValidationError) as e:
+        LOG.warning('Database lookup for action with id="%s" resulted in '
+                    'exception: %s', action_id, e)
+        raise StackStormDBObjectNotFoundError('Unable to find action with '
+                                              'id="%s"' % action_id)
+
+    return action
+
+
+def create_execution_request_from_liveaction(liveaction, delay=None,
+                                             priority=None, affinity=None):
+    """
+        Create execution request from liveaction.
+    """
+    execution_request = ExecutionQueueDB()
+    execution_request.liveaction = liveaction.to_serializable_dict()
+    execution_request.start_timestamp = date.append_milliseconds_to_time(
+        liveaction.start_timestamp,
+        delay or 0
+    )
+    execution_request.delay = delay
+    execution_request.priority = priority
+    execution_request.affinity = affinity
+
+    return execution_request
+
+
+@coordination.Locker('st2schedulerqueue')
+def pop_next_execution():
+    """
+        Sort executions by fifo and priority and get the latest, highest priority
+        item from the queue and pop it off.
+    """
+    query = {
+        "start_timestamp__lte": date.get_datetime_utc_now(),
+        "order_by": [
+            "-priority",
+            "start_timestamp",
+            "delay",
+        ]
+    }
```

Comment:
This query doesn't seem to calculate if a delay wait is over.

Reply:
I need to take delay out of that query. It was way more efficient to store the query start time as live action start + delay since I can make the one query based on anything with start timer <= current time.

Reply:
I agree - storing absolute time is better, imo.

This way we can utilize it for any executions which need to be delayed and executed at some later point (also for retry policy - https://github.com/StackStorm/st2/pull/4397#issuecomment-432176089, etc.)

Reply:
Yeah. My point here is that this query as it stands right now is not actually comparing current time, start time, and delay. As far as I can tell, the query find any records older then current time and pick the first one.  Where's the delay calculation?

Reply:
The only thing that needs to happen here is removal of delay, as I said. This query will sort by priority first, then by timestamp. So any available executions will be executed first by priority then by first added. Sorting by delay was left over from my originally trying to keep start_timestamp the same as the liveaction, but I abandoned that because it's impossible to query for that efficiently.

Reply:
Yeah, this is kinda confusing at first glance. I see you calculated the delay into the start time when you created this record.  To make this clear, I would rename `start_timestamp` to `scheduled_start_timestamp` so not to be confused with actual start timestamp of the live action and you already say you will rename `delay` from the table and from the query.

Reply:
Yeah, I agree. Am planning to change it. Left over artifact from how I was calculating it prior.